### DISCUSSION TITLE
GSOC 2023 - Visualization API with Matplotlib

### DISCUSSION THREAD

**Discussion by ytongwu** [networkx/networkx#6559] [General]
Hi, 
I'm Yutong and I'm new to NetworkX and open source. I'm interested in the project "Visualization API with Matplotlib" and would like to contribute! I have installed NetworkX and familiarized myself with the documentations. During the installation, there were some warnings after running the tests and I passed 5044 tests and skipped several ones. I'm wondering if the warnings would affect later contribution.
@dschult Could you point me to some possible contributions I could make and the parts of API that need improvement? Would really appreciate any advice on how to start working on the project and available learning resources. Thank you!

**Comment by dschult** [comment]
The `drawing/nx_pylab.py` module holds most of the interface for drawing within NetworkX. The tests for it are in `drawing/tests/test_pylab.py`. You will see that there are a very large number of keyword arguments, many of which are passed along to the matplotlib "Artists" that actually do the rendering. Some of the keyword arguments accept scalars or arrays/lists, some only arrays and some only scalars. One goal would be to figure out whether things have changed in matplotlib enough to standardize these features to be e.g. all scalars or arrays.  You will also see that the edges are drawn in two different ways depending on properties of the inputs and of the graph. We use a LineCollection when we can because of its speed. But edges are often more than just lines. So we have another set of code to draw edges using lists of FancyArrowPatch? We use this whenever the edge is bent, we have arrows, and a few other conditions including some input parameters.  It seems like that part of the API could be reworked. 

Also, there is a PR for the repository about drawing multiedges by bending the edges by different amounts depending on how many edges are between the two nodes. And there are probably many Issues and PRs (some of which might be closed) where useful discussions have taken place about how to design the interface.  In addition, there is a second repository networkx/GraVe which implements a propotype visualization library for NetworkX. That has been dormant for a few years and I'm not sure if it even runs with current versions.  These are sources of information, history and code about the API.

Often some easy PRs to get started involve doc_string improvements and/or testing improvements. As you learn about the drawing features, try to identify any parts that are not clear for you and if they are easy to improve, go ahead and make a PR.  You can also look through the current issues (and PRs) to see if there is a fairly simple one you could address.
:)

**Reply by ytongwu** [reply]
Thank you so much for the detailed explanation! I will definitely try to get a better understanding and fix some issues to make contributions.

**Reply by ytongwu** [reply]
Hi @dschult ,
I have gone through the resources you provided and they are really helpful! 

For the `drawing/nx_pylab.py`, I'm not sure if I understand you correctly, do you suggest that we could unify all the other one-type inputs as scalars and all sequences as lists, or modify the parameters that accept both scalars and arrays to only accept one of them? I noticed that the docstrings of the keyword arguments vary throughout the file: some use array, others use list or sequence, so one way could be to unify these expressions into like 'scalars or sequence'. I'm also wondering if there are other parts that I can look into to get a better understanding about what I could do in the project.

The PR of drawing bending multiedges is very interesting and I also saw some discussions about new designs and compatibility with Matplotlib. I'm also trying to come up with possible new features to ask, but it may take some time lol. I just created a PR for docstring improvement in the nx_pylab.py, and I'm curious if there are changes made in this file, are the API and pdf version going to be updated as well? Thank you for your time!
